### PR TITLE
[no-object-literal-type-assertion] New option: "allow-arguments"

### PR DESCRIPTION
**Pull request by karol-majewski**
#### PR checklist

- [x] Addresses an existing issue: #4333
- [x] Enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:

This option allows using a type assertion when an object literal is used as an argument. A common use case is the second parameter for `Array.prototype.reduce`, where an empty object is often used as the initial value.

#### CHANGELOG.md entry:

[new-rule-option] [`no-object-literal-type-assertion`](https://palantir.github.io/tslint/rules/no-object-literal-type-assertion/) rule accepts an option called `allow-arguments`. It allows using type assertions on object literals used inside call expressions.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @karol-majewski! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: src/rules/noObjectLiteralTypeAssertionRule.ts [hunk 1]
```diff
@@ -37,23 +45,35 @@ export class Rule extends Lint.Rules.AbstractRule {
             The compiler will warn for excess properties with this syntax, but not missing required fields.
             For example: \`const x: { foo: number } = {}\` will fail to compile, but
             \`const x = {} as { foo: number }\` will succeed.`,
-        optionsDescription: "Not configurable.",
-        options: null,
-        optionExamples: [true],
+        optionsDescription: Lint.Utils.dedent`
+            One argument may be optionally provided:
+
+            * \`${OPTION_ALLOW_ARGUMENTS}\` allows type assertions to be used on object literals inside call expressions.`,
```
Comment (JoshuaKGoldberg) [inline src/rules/noObjectLiteralTypeAssertionRule.ts]: Although TSLint has traditionally had rule arguments provided as strings, that's starting to be a bit of a pain when they become more complex. Let's have the rule take in an object:

```ts
[true, {
    "allow-arguments": true
}]
```

File: test/rules/no-object-literal-type-assertion/allow-arguments/test.ts.lint [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Allow object literals to be asserted when used as arguments.
```
Comment (JoshuaKGoldberg) [inline test/rules/no-object-literal-type-assertion/allow-arguments/test.ts.lint]: Just to make sure the `allow-arguments` option doesn't completely break the rule in the future, could you copy & paste the existing test cases into this file? They should still cause complaints for non-argument assertions.

File: src/rules/noObjectLiteralTypeAssertionRule.ts (hunk 1 above)
Comment (karol-majewski) [reply src/rules/noObjectLiteralTypeAssertionRule.ts]: Done 👍 I borrowed the new schema from the `noImportSideEffectRule.ts` rule.

https://github.com/palantir/tslint/blob/9785477f26c1e0f926a5a34ba9170f1784a728c7/src/rules/noImportSideEffectRule.ts#L29-L41

File: test/rules/no-object-literal-type-assertion/allow-arguments/test.ts.lint (hunk 2 above)
Comment (karol-majewski) [reply test/rules/no-object-literal-type-assertion/allow-arguments/test.ts.lint]: Of course. Done 👍

**Comment by Luixo** [comment]
@adidahiya 
Hey! I couldn't find any info about the review routine in palantir and I hope it's not a very rude question to ask, but is there any evaluated timing for the review? :)

File: src/rules/noObjectLiteralTypeAssertionRule.ts [hunk 3]
```diff
@@ -37,23 +49,44 @@ export class Rule extends Lint.Rules.AbstractRule {
...
             \`const x = {} as { foo: number }\` will succeed.`,
-        optionsDescription: "Not configurable.",
-        options: null,
-        optionExamples: [true],
+        optionsDescription: Lint.Utils.dedent`
+            One option may be configured:
+
+            * \`${OPTION_ALLOW_ARGUMENTS}\` allows type assertions to be used on object literals inside call expressions.`,
+        options: {
+            type: "array",
```
Comment (adidahiya) [inline src/rules/noObjectLiteralTypeAssertionRule.ts]: you don't need to wrap the object in an array, you can just bring the object out as the top level `options` value

Comment (karol-majewski) [reply src/rules/noObjectLiteralTypeAssertionRule.ts]: Done! Thank you.

**Pull request merged**